### PR TITLE
path: add tests for minimum spanning tree functions

### PR DESCRIPTION
**Pull request by kortschak**
* Replace incorrect Prim's implementation.
* Make functions return weight.
* Drop default to uniform cost function since this is essentially meaningless here.

@vladimir-ch Please take a look.

Note that I had a map-free implementation that this is derived from, but again the acounting needs to be done on a modified node and so would require a copy of the entire graph. This is easier.

If you would like me to add a check for the actual structure of the MST, I'll add that.

### CODE REVIEW

**Comment by kortschak** [comment]
@vladimir-ch I have differentiated Prim and Kruskal with documentation explaining that Prim will only return a single tree of the forest when the input graph is not connected while Kruskal will return the forest and sum of weights. We can make them both behave like Kruskal with a minor change to the logic (detect when the min key is inf and restart from there). What do you think, as is or with that addition?

**Comment by vladimir-ch** [comment]
Give me some time.

**Comment by kortschak** [comment]
I've realised there is a subtle error in Prim and I am fixing that. I will go with the MSF approach.

**Comment by kortschak** [comment]
It's ready, but take your time. Sorry for the flux.

File: path/spanning_tree.go [hunk 1]
```diff
@@ -5,101 +5,166 @@
...
 	"sort"
 
 	"github.com/gonum/graph"
 	"github.com/gonum/graph/concrete"
-	"github.com/gonum/graph/internal"
 )
 
-// EdgeListerGraph is an undirected graph than returns its complete set of edges.
-type EdgeListerGraph interface {
+// UndirectedWeighter is an undirected graph that returns distinct edge weights.
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: I don't understand the comment well because of the word "distinct". Distinct from what?

File: path/spanning_tree.go [hunk 2]
```diff
@@ -5,101 +5,166 @@
...
 
 // Prim generates a minimum spanning tree of g by greedy tree extension, placing
-// the result in the destination. The destination is not cleared first.
-func Prim(dst graph.MutableUndirected, g EdgeListerGraph) {
-	var weight Weighting
-	if wg, ok := g.(graph.Weighter); ok {
-		weight = wg.Weight
-	} else {
-		weight = UniformCost(g)
+// the result in the destination. The destination is not cleared first. The weight
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: I would say `... in the destination dst. dst is not cleared first. ...`

File: path/spanning_tree.go [hunk 3]
```diff
@@ -5,101 +5,166 @@
...
+	nodes := g.Nodes()
+	if len(nodes) == 0 {
+		return 0
 	}
-
-	nlist := g.Nodes()
-
-	if nlist == nil || len(nlist) == 0 {
-		return
+	u := nodes[0]
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: Move this line before line 40?

File: path/spanning_tree.go [hunk 4]
```diff
@@ -5,101 +5,166 @@
...
-		dst.SetEdge(min)
-		remain.Remove(min.From().ID())
 	}
+	return w
+}
+
+// primQueue is an Prim's priority queue.
+type primQueue struct {
+	indexOf map[int]int
+	nodes   []concrete.Edge
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: `Prim()`'s code is almost self-documenting, but it took me time to understand the (key) role of the queue. Even though not exported, I think it would be better to document it a bit more and avoid decyphering the code. What took me most time to sort out is what is this slice (it is called `nodes` but it is a slice of edges), what edges are stored here and what are their From and To nodes (`update()` is called with To, From, but interprets them as From, To).
*

File: path/spanning_tree.go [hunk 5]
```diff
@@ -5,101 +5,166 @@
...
-
-		sort.Sort(byWeight(edges))
-		min := edges[0]
-		dst.SetEdge(min)
-		remain.Remove(min.From().ID())
 	}
+	return w
+}
+
+// primQueue is an Prim's priority queue.
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: an -> a

File: path/spanning_tree.go [hunk 6]
```diff
@@ -5,101 +5,166 @@
...
+}
+
+func (q *primQueue) Pop() interface{} {
+	n := q.nodes[len(q.nodes)-1]
+	q.nodes = q.nodes[:len(q.nodes)-1]
+	delete(q.indexOf, n.From().ID())
+	return n
+}
+
+// key returns the key for the node n and whether the node is
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: ... for the node u ...

File: path/spanning_tree.go [hunk 7]
```diff
@@ -5,101 +5,166 @@
...
+	if !ok {
+		return math.Inf(1), false
 	}
+	return q.nodes[i].Weight(), ok
+}
 
-	edgeList := g.Edges()
-	edges := make([]concrete.Edge, 0, len(edgeList))
-	for _, e := range edgeList {
+// update updates the node's position in the queue with the new key.
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: Another point of confusion, the queue holds edges but this comment talks about node's position. This comment should be improved.

File: path/spanning_tree.go [hunk 8]
```diff
@@ -5,101 +5,166 @@
...
+	i, ok := q.indexOf[id]
+	if !ok {
+		return
+	}
+	q.nodes[i].T = v
+	q.nodes[i].W = key
+	heap.Fix(q, i)
+}
+
+// UndirectedWeightLister is an undirected graph that returns distinct edge weights and
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: If you edit the "distinct" comment above, edit it also here.

File: path/spanning_tree.go [hunk 9]
```diff
@@ -5,101 +5,166 @@
...
+// the set of edges in the graph.
+type UndirectedWeightLister interface {
+	UndirectedWeighter
+	Edges() []graph.Edge
+}
+
+// Kruskal generates a minimum spanning tree of g by greedy tree coalescence, placing
+// the result in the destination. The destination is not cleared first. The weight
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: Destination dst also here.

File: path/spanning_tree.go (hunk 1 above)
Comment (kortschak) [inline path/spanning_tree.go]: This is the terminology used in the algorithm proofs. Distinct means there are no repeated weights in the set of E. The consequence of this invariant being broken is not terrible (the algorithms find *an* MST instead of *the* MST), so I don't check for it - the user should know whether they care about the consequence and what it means.

File: path/spanning_tree.go (hunk 2 above)
Comment (kortschak) [inline path/spanning_tree.go]: I'll meet you half way: "... in the destination, dst. The destination is not cleared first." I don't like "dst. dst".

File: path/spanning_tree.go (hunk 3 above)
Comment (kortschak) [inline path/spanning_tree.go]: Done, with a flourish.

File: path/spanning_tree.go (hunk 4 above)
Comment (kortschak) [inline path/spanning_tree.go]: Mmmm. It's actually a funny bit of code. The naive way to implement the algorithm is along the line of how Jeff did it, with a sorted list of edges. The optimised forms use a heap (binary or Fibonacci) of nodes weighted by the minimum weighted out edge. This code does that, so the heap is actually a heap of From nodes - originally I defined a primNode, but it was isomorphic with an Edge, so I dropped the type and used concrete.Edge instead.

This is also what contributes to the weirdness with the edge directions; we are looking for edges that lead *to* our query and minimising that node.

Maybe the thing to do is make a primNode derived from a concrete.Edge?

File: path/spanning_tree.go (hunk 6 above)
Comment (kortschak) [inline path/spanning_tree.go]: Done.

File: path/spanning_tree.go (hunk 7 above)
Comment (kortschak) [inline path/spanning_tree.go]: I think this is answered above.

File: path/spanning_tree.go (hunk 9 above)
Comment (kortschak) [inline path/spanning_tree.go]: As above.

**Comment by kortschak** [comment]
PTAL

File: path/spanning_tree.go (hunk 1 above)
Comment (vladimir-ch) [inline path/spanning_tree.go]: > This is the terminology used in the algorithm proofs. Distinct means there are no repeated weights in the set of E.

Ok, now I understand.

> The consequence of this invariant being broken is not terrible (the algorithms find an MST instead of the MST), so I don't check for it - the user should know whether they care about the consequence and what it means.

Are the consequences so obvious that it is not necessary to mention them? By speaking about "distinct edge weights" we show that repeated edge weights are somehow important but we don't document how and why.

Also, I must ask if Prim's and Kruskal's algorithms, by being greedy, find the minimum spanning tree instead of just a spanning tree and if we should say something in respect in the docs.

File: path/spanning_tree.go (hunk 4 above)
Comment (vladimir-ch) [inline path/spanning_tree.go]: I could sense that the use of heap is a kind of optimization of the naive implementation. I like it and I think that primNode is probably not necessary. Just sprinkling `primQueue` and `update()` with some explaining comments should be enough, but I leave it up to you.

File: path/spanning_tree.go (hunk 1 above)
Comment (kortschak) [inline path/spanning_tree.go]: I guess it depends on how much hand holding we should be doing. My view (not necessarily the correct view) is that people should be able to look up when the algorithms are correct to use. Though I could clarify in Prim and Kruskal that they find "a minimum spanning tree, which will be the minumum spanning tree if the edge weights are distinct" and drop the distinct part from the interface doc comment.

Comment (vladimir-ch) [inline path/spanning_tree.go]: > Though I could clarify in Prim and Kruskal that they find "a minimum spanning tree, which will be the minumum spanning tree if the edge weights are distinct" and drop the distinct part from the interface doc comment.

The wording you propose is nice, not much of a hand-holding but still transmits the key message.

Comment (kortschak) [inline path/spanning_tree.go]: That wording didn't work with the structure of the rest of the para, but I have something close in there now.

File: path/spanning_tree.go [hunk 10]
```diff
@@ -5,101 +5,173 @@
...
+// as +Inf.
+func (q *primQueue) key(u graph.Node) (key float64, ok bool) {
+	i, ok := q.indexOf[u.ID()]
+	if !ok {
+		return math.Inf(1), false
+	}
+	return q.nodes[i].Weight(), ok
+}
+
+// update updates the node's position in the queue with the new key.
```
Comment (vladimir-ch) [inline path/spanning_tree.go]: May I have one last request to update this comment to say which node's position is updated (because two are given) and what is the role of the other one?

**Comment by vladimir-ch** [comment]
After addressing the one last request LGTM

File: path/spanning_tree.go (hunk 10 above)
Comment (kortschak) [inline path/spanning_tree.go]: Done.
